### PR TITLE
Allow using labels as argument to `rst`

### LINKED ISSUE #448: [Feature request] Allow RGBLINK to determine the target of a `rst`
This shouldn't be too problematic: check that the symbol resolves to one of the 8 possible locations, and produce the corresponding opcode.
This would be part of a new object file specification, breaking backwards compatibility, but the new linker has support for reading several versions, so I believe it should be OK.

### PR DESCRIPTION
**Pull request by ISSOtm**
Fixes #448.

This is also a test regarding an object file version bump—the next RGBDS release will use the RGB7 format, whose spec will change during its development.
An attempt has been made in RGBLINK to keep handling the RGB6 format, but the way it's currently implemented has a flaw: RGB6 files can contain the new RPN opcode, and RGBLINK won't complain. In practice, a well-formed RGB6 file will not contain such an opcode, and supporting an extra one shouldn't be doing any harm.
On the other hand, duplicating more of the object-processing code (I have copy-pasted the function because it is likely to change during development of RGB7, especially due to #292) depending on version sounds risky, and I'd rather avoid it.
Thoughts?

### CODE REVIEW

**Comment by aaaaaa123456789** [comment]
Please avoid changing the spec later without changing the version number. That's almost guaranteed to result in issues building.

**Comment by AntonioND** [comment]
Personally I think there is no point in supporting old object file formats. We haven't supported older formats for years and nobody has complained. The only reason why you would want to do that is to have a closed source library or something like that, which very niche. Considering how niche the toolchain itself is... I think it's just extra code to maintain for no real benefit.

**Comment by ISSOtm** [comment]
> Please avoid changing the spec later without changing the version number. That's almost guaranteed to result in issues building.

That's not supposed to be an issue, as master is unstable with no real releases; the spec would be considered frozen on the next release (which would include updating of https://rednex.github.io/rgbds/rgbds.5.html)

> Personally I think there is no point in supporting old object file formats. We haven't supported older formats for years and nobody has complained. The only reason why you would want to do that is to have a closed source library or something like that, which very niche. Considering how niche the toolchain itself is... I think it's just extra code to maintain for no real benefit.

My rationale for that one was that several tools appear to output RGBDS object files directly, and maybe nobody has complained because RGB6 has been there for a while? (That's how it seems to me, at least). I believe it also makes sense to support the *previous* version if the added cost is small, so people may have an easier time transitioning between versions.

**Comment by aaaaaa123456789** [comment]
> That's not supposed to be an issue, as master is unstable with no real releases; the spec would be considered frozen on the next release

You know very well this is not the case. People use `master` all the time, for various good reasons. It's not a good idea to break people's builds because you don't want to make a branch.  
If you really need to iterate over the new format (which I doubt — if you already know what it needs, why not define it properly at once?), then _add a minor version field_, so the iteration doesn't break people's builds. Being forced to rebuild a .o is not a problem. Having the tool crash or output the wrong code (or even worse, go completely haywire) because it's using a different spec to read the .o file even though it matches the expected version number, on the other hand, is a big one.

I don't understand why you refuse to acknowledge this problem, even though you're well aware of it. `master` has never been unstable here, not for as long as I've been using the tool at least. Even if your approach is right in theory (which I doubt), in practice, that doesn't matter. People's builds will break if you do this.

> My rationale for that one was that several tools appear to output RGBDS object files directly, and maybe nobody has complained because RGB6 has been there for a while? (That's how it seems to me, at least). I believe it also makes sense to support the _previous_ version if the added cost is small, so people may have an easier time transitioning between versions.

If this is the case, and the complexity becomes too much to handle, perhaps a tool to convert to the newest object file format version would solve the problem? I've never heard of tools other than `rgbasm` outputting object files, though.

**Comment by ISSOtm** [comment]
There is no point in making a new branch for this, as people will just switch to it. 
I believe that users of `master` should be competent enough to know that they are on an unstable track—that's why breakage happens often—and be prepared for some breakage as well. Having to care about every single commit being sound and safe is not feasible (at least not with the current CI), and that's why I suggested making pre-releases, but that idea went nowhere.

To put it simply: RGB7 isn't defined yet. I Do not know yet how it will be designed, both because I don't know which features will be changed, and because I can't plan in advance how they will be implemented, especially in regards to changes required by bug fixes.

Adding a minor version field is basically extending the version number, but it does not solve the problem because it's not automatically increased. Having it be modified by two different PRs at once will happen and adds extra required attention I do not want to have to give for the sake of an unstable track.

I know for a fact that GBTD / GBMB allow outputting .o files. I have no usage statistics on anything though.

**Comment by AntonioND** [comment]
> There is no point in making a new branch for this, as people will just switch to it.
> I believe that users of `master` should be competent enough to know that they are on an unstable track—that's why breakage happens often—and be prepared for some breakage as well. Having to care about every single commit being sound and safe is not feasible (at least not with the current CI), and that's why I suggested making pre-releases, but that idea went nowhere.

Exactly. If you use non-tagged versions of a tool, expect changes every now and then. That's no reason to break things on purpose on master, but nobody should expect master to be completely stable, and iterations over a complicated design look to me like a valid reason to break master.

**Comment by aaaaaa123456789** [comment]
Breaking changes are generally OK. If you're using `master`, you'd sometimes expect it to fail and give you an error because a feature changes. That's fine.

This is not just giving errors. This can _result in the wrong code being generated_ because a pre-existing object file is being read in a different way. The user would have no indication of this, even.

A minor version number solves this problem, causing builds to fail as expected. (Builds failing because the tool updated are generally OK — you just clean and rebuild.) As for the PR issue, it should be easy to check that the version number has been changed in it, and fix conflicts when fixing other merge conflicts.

Or you can just use the commit hash as the minor version number.

**Comment by AntonioND** [comment]
> Or you can just use the commit hash as the minor version number.

Relying on the git repository to be present isn't a great idea.

In any case, I don't see the benefit of having a minor version number. In practice, the check is to see if the version is exactly the same. Changing the major version number achieves the same effect. Also, we don't really need to worry about running out of space for version numbers in the header...

Personally I think that changing the major number between releases is fine, and that the format should only be locked when a release is tagged. That is, any change after a tagged release implies an increase in the version number of the object file, but you only need to do it once per release.

Also, I think that it is nice to tag a release whenever there are big changes, and modifying the format of the object files counts to me as a big change. Maybe not right away, but after 1-2 weeks.

**Comment by aaaaaa123456789** [comment]
It seems to me that we're mostly in agreement regarding how to handle the version number, except for the major/minor split.

The issue is that the current version number can't grow past 9. Do you have any better ideas on how to extend it?

**Comment by ISSOtm** [comment]
> The user would have no indication of this, even.

It's possible to add a warning message to RGBLINK to warn users of the unstable object file spec.

> The issue is that the current version number can't grow past 9. Do you have any better ideas on how to extend it?

What I currently have in mind is to add an "extended" version number when RGB9 is reached; however, that will be in a while, as we still have RGB7 and RGB8 before that, which should be a couple years.

> As for the PR issue, it should be easy to check that the version number has been changed in it, and fix conflicts when fixing other merge conflicts.

My point is when two PRs try to increase the version number to the same value, which will *not* cause a merge conflict, and would instead require special attention, whenever merging one, to change the number in all others. That means having several such PRs would be tedious, which is either more involvement (and I don't want to have that added) or only one such PR at a time (slowing down development further).

**Comment by aaaaaa123456789** [comment]
> It's possible to add a warning message to RGBLINK to warn users of the unstable object file spec.

Such a warning is non-actionable and not particularly helpful, since it would appear all the time. It's the kind of disclaimer you write to cover your ass when a complaint happens, not to be helpful.

> ...as we still have RGB7 and RGB8 before that, which should be a couple years.

If we go by this idea:

> Also, I think that it is nice to tag a release whenever there are big changes, and modifying the format of the object files counts to me as a big change. Maybe not right away, but after 1-2 weeks.

...then merging this PR means that a new release is imminent, and RGB7 will be locked in by approx. the end of the month. (This is an acceptable compromise IMO.) Since you have more changes queued, I sincerely doubt two version numbers will last a couple of _months_, let alone years.

> My point is when two PRs try to increase the version number to the same value, which will not cause a merge conflict, and would instead require special attention, whenever merging one, to change the number in all others.

You only have to check the version numbers prior to merging a PR, and only increase it in that particular PR before merging if it has fallen out of date.

**Comment by aaaaaa123456789** [comment]
Just to make myself clear: I'm not against making multiple changes while merging a series of PRs over the course of several days, nor against making a quick fix a day or two after merging said PRs. But that's not what you propose.

Your proposal consists of the format changing multiple times over an undefined, lengthy period of time, all without updating the version number. This is almost guaranteed to cause issues and it defeats the purpose of having a versioned file format in the first place. All because of the technical debt of having a single-digit version number?

**Comment by NieDzejkob** [comment]
> My point is when two PRs try to increase the version number to the same value, which will _not_ cause a merge conflict, and would instead require special attention, whenever merging one, to change the number in all others.

Possible solution: make pre-release version numbers be the number of the most-recently merged PR.

**Comment by ISSOtm** [comment]
Here is a proposition for an usable pre-release system:
- Every so often (two? three months?) a pre-release is made.
- A revision number is added to RGB7 onwards. This number is incremented after each prerelease which broke the file format, such that object files generated between incompatible prereleases are not accepted
- The actual file format number (RGBx) is only incremented after stable releases, resetting the revision count to 0

**Comment by AntonioND** [comment]
I don't see the point in this:

> The actual file format number (RGBx) is only incremented after stable releases, resetting the revision count to 0

If we are going to extend the version number and add a "revision" we might as well make the revision the actual version number. If there is a version or revision mismatch, the linker fails. If the revision number is a 32-bit number, we have an essentially infinite number of available revisions.

**Comment by aaaaaa123456789** [comment]
I agree; you might as well hop to `RGB9` (to prevent further increases) and use the new field as the 'true' number.

**Comment by ISSOtm** [comment]
The reason why I wanted to keep the current version number was to have a major number in addition to the revision number, which would act more like a minor. The idea was to make it clearer which versions are stabilized (for external tools to optionally write and parse such objects)

**Comment by AntonioND** [comment]
The thing is, semantic versioning doesn't make sense if you increase the major version number of the object file based on the release of the toolchain. Basically, any minor release of RGBDS would trigger a change of the major version of the object files. In short, you can't assume that a major version change makes your tool incompatible, and you can't assume that a minor version change leaves your tool as compatible. You might as well just have a unique version number at that point.

If you want, you can always follow the pattern of ``version = version_major << 16 | version_minor;``.

**Comment by ISSOtm** [comment]
I've done that. If everyone is happy with this, I will merge this and start working on other modifications to the object format.

File: src/rgbds.5 [hunk 1]
```diff
@@ -33,7 +33,7 @@ is a 0‐terminated string of
 .Bd -literal
 ; Header
 
-BYTE    ID[4]            ; "RGB6"
+BYTE    ID[4]            ; "RGB7"
```
Comment (AntonioND) [inline src/rgbds.5]: Update this to say ``"RGB9"`` and add the new ``revision`` field.

**Comment by ISSOtm** [comment]
Awesome, thank you! I will start working on fixing `@` expressions being incorrectly computed by RGBLINK, which requires passing the value of PC in expressions.

**Pull request merged**